pdp/storage.go

```diff
@@ -57,6 +57,101 @@ func (s *PolicyStorage) NewTransaction(tag *uuid.UUID) (*PolicyStorageTransactio
 	return &PolicyStorageTransaction{tag: *tag, attrs: s.attrs, policies: s.policies}, nil
 }
 
+// GetPath returns Policy/PolicySet at path if it exists, otherwise return nil, err
+func (s *PolicyStorage) GetPath(path []string) (Iterable, error) {
+	var (
+		err error
+		// [PolicySet, Policy, Rule] should always be iterable
+		iter = s.policies.(Iterable)
+	)
+	if rootID, ok := s.policies.GetID(); len(path) == 0 || !ok || rootID != path[0] {
+		return nil, fmt.Errorf("Invalid root id or hidden root")
+	}
+	for _, id := range path[1:] {
+		iter, err = iter.FindNext(id)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return iter, nil
+}
+
+// GetSubtree obtains subtree of iter within depth in JSON format
+func GetSubtree(iter Iterable, depth uint) string {
+	if id, ok := iter.GetID(); ok {
+		idInfo := fmt.Sprintf("\"id\":%s", strconv.Quote(id))
+		if depth > 0 {
+			nChild := iter.NextSize()
+			cIdx := 0
+			children := make([]string, nChild)
+			for i := 0; i < nChild; i++ {
+				child := GetSubtree(iter.GetNext(i), depth-1)
+				if len(child) > 0 {
+					children[cIdx] = child
+					cIdx++
+				}
+			}
+			if cIdx > 0 {
+				return fmt.Sprintf("{%s,\"elems\":[%s]}", idInfo,
+					strings.Join(children[:cIdx], ","))
+			}
+		} else if iter.NextSize() > 0 {
+			return fmt.Sprintf("{%s,\"elems\":\"...\"}", idInfo)
+		}
+		return fmt.Sprintf("{%s}", idInfo)
+	}
+	return ""
+}
```

Comment:
I know GetSubtree returning JSON string is weird, but it seems like the cheapest and simplest way to limit the height of the tree for our purpose. The recursive depth is min(depth, height of the tree). An alternative way is to create temporary iterables (as to not change the storage tree), then serialize it in our query API on pdpserver. This is more generalized and reusable later.

Reply:
That's why iterable isn't good. Recursive marshaling can easily limit depth by passing current depth to each marshal function.

Reply:
Anyway it should be rewritten to not return string of arbitrary size. Instead marshallers  should be able to use fixed size buffer and send the buffer to output as soon as it's full. It can be done with callback or channel approach. Latest is more idiomatic while the callback seems to be more performant.